### PR TITLE
[Issue 645] Restructure error response format

### LINKED ISSUE #645: [Task]: API Error Handling
### Summary

There are a number of different ways to display errors for requests that have the wrong field type, validation, or other errors that aren't covered by response codes.

Our docs say that errors get returned like below, but we don't currently use this error list. We want to correct our responses to have a consistent format for errors:
```json
{
  "errors": [
    {
      "field": "string",
      "message": "string",
      "type": "string",
      "value": "string"
    }
  ],
  "message": "some sort of error message",
  "status_code": 422,
}
```

#### Current State

Request validation errors are structured like:

```json
{
  "detail": {
    "json": {
      "category": [
        "Must be one of: D, M, C, E, O."
      ],
      "sorting": {
        "order_by": [
          "Must be one of: opportunity_id, agency, opportunity_number, created_at, updated_at."
        ]
      }
    }
  },
  "message": "Validation error"
}
```
Which is a set of nested objects in the detail object, and only contains messages from our underlying Marshmallow validation. While the validation is good, its structure isn't quite what we want.

### Acceptance criteria

- [ ] Marshmallow error handling structured to return like our docs indicate

### PR DESCRIPTION
**Pull request by chouinar**
## Summary
Fixes #645

## Changes proposed
Validation errors from Marshmallow are now formatted into the errors list on the response object.

## Context for reviewers
I know this is a lot, it is a bit difficult to split up, and I'm happy to walk you through it.

There are a few high-level fundamental changes:
1. Each of the Marshmallow schema + fields + validators needed to be restructured a little to allow for validation error type + a message.
2. An error processor was attached to the application to restructure the Marshmallow errors
3. All error scenarios should go through the error processor and be re-formatted if necessary (mainly just Marshmallow at the moment).

## Additional information
A lot of tests focused on testing these changes, including a few test-only endpoints that validate various error

The endpoints still work properly locally, and more-or-less are unmodified outside of a few small edge cases that the new field classes now account for (nullability).

### CODE REVIEW

File: api/openapi.generated.yml [hunk 1]
```diff
@@ -27,25 +27,28 @@ paths:
                     $ref: '#/components/schemas/Healthcheck'
                   status_code:
                     type: integer
+                    minimum: -2147483648
+                    maximum: 2147483647
```
Comment (chouinar) [inline api/openapi.generated.yml:31]: I might want to remove what added this - this was porting over something that by default limits integers to the values the Postgres DB supports, but it might be more confusing to have this on fields like status code

**Review by acouch — APPROVED** [review]
